Sherlock/yoda/sfinae.h

```diff
@@ -83,6 +94,41 @@ constexpr auto HasOperatorLess(int)
   return true;
 }
 
+template <typename T_KEY, bool IS_ENUM = false>
+struct GenericHasStdHash {
+  static constexpr bool value() { return HasStdHash<T_KEY>(0); }
+};
+
+template <typename T_KEY>
+struct GenericHasStdHash<T_KEY, true> {
+  static constexpr bool value() { return EnumHasStdHash<T_KEY>(0); }
+};
+
+template <typename T_KEY, bool HAS_CUSTOM_HASH_FUNCTION, bool IS_ENUM>
+struct T_HASH_SELECTOR;
+
+template <typename T_KEY>
+struct T_HASH_SELECTOR<T_KEY, true, false> {
+  struct HashFunction {
+    size_t operator()(const T_KEY& key) const { return static_cast<size_t>(key.Hash()); }
+  };
+  typedef HashFunction type;
+};
+
+template <typename T_KEY>
+struct T_HASH_SELECTOR<T_KEY, false, true> {
+  struct HashFunction {
+    typedef typename std::underlying_type<T_KEY>::type UT;
+    size_t operator()(const T_KEY& key) const { return std::hash<UT>()(static_cast<UT>(key)); }
+  };
+  typedef HashFunction type;
+};
+
+template <typename T_KEY>
+struct T_HASH_SELECTOR<T_KEY, false, false> {
+  typedef std::hash<T_KEY> type;
+};
```

Comment:
This SFINAE LGTM++.

Perhaps a few simple `static_assert()`-s are worth keeping?